meteor/client/lib/rundown.ts

```diff
@@ -219,6 +221,68 @@ export namespace RundownUtils {
 			.replace(/_/g, '-')
 	}
 
+	export function getSegmentsWithPartInstances(
+		playlist: RundownPlaylist,
+		segmentsQuery?: Mongo.Query<DBSegment> | Mongo.QueryWithModifiers<DBSegment>,
+		partsQuery?: Mongo.Query<DBPart> | Mongo.QueryWithModifiers<DBPart>,
+		partInstancesQuery?: Mongo.Query<PartInstance>,
+		segmentsOptions?: FindOptions<DBSegment>,
+		partsOptions?: FindOptions<DBPart>,
+		partInstancesOptions?: FindOptions<PartInstance>
+	): Array<{ segment: Segment; partInstances: PartInstance[] }> {
+		const { segments, parts: rawParts } = playlist.getSegmentsAndPartsSync(
+			segmentsQuery,
+			partsQuery,
+			segmentsOptions,
+			partsOptions
+		)
+		const rawPartInstances = playlist.getActivePartInstances(partInstancesQuery, partInstancesOptions)
+
+		const partsBySegment = _.groupBy(rawParts, (p) => p.segmentId)
+		const partInstancesBySegment = _.groupBy(rawPartInstances, (p) => p.segmentId)
+
+		return segments.map((segment) => {
+			const segmentParts = partsBySegment[unprotectString(segment._id)] || []
+			const segmentPartInstances = partInstancesBySegment[unprotectString(segment._id)] || []
+
+			if (segmentPartInstances.length === 0) {
+				return {
+					segment,
+					partInstances: segmentParts.map(wrapPartToTemporaryInstance),
+				}
+			} else if (segmentParts.length === 0) {
+				return {
+					segment,
+					partInstances: _.sortBy(segmentPartInstances, (p) => p.part._rank),
+				}
+			} else {
+				const partInstanceMap = new Map<PartId, PartInstance>()
+				for (const part of segmentParts) partInstanceMap.set(part._id, wrapPartToTemporaryInstance(part))
+				for (const partInstance of segmentPartInstances) {
+					if (
+						partInstance._id === playlist.nextPartInstanceId ||
+						partInstance._id === playlist.previousPartInstanceId
+					) {
+						const currentValue = partInstanceMap.get(partInstance.part._id)
+						if (currentValue?._id === playlist.currentPartInstanceId) {
+							// Show the current instead of the next or previous PartInstance if they are the same
+							continue
```

Comment:
I don't exactly understand this part? 🤔 Why would the Part Instance (of the same ID) be listed multiple times in `segmentPartInstances`? Wouldn't there be a single instance of this Part Instance? How can there be more then one? Shouldn't this check if the `partId` of these Part Instances (current/next/previous) (and not their IDs) is the same, and then _show the current instead of the next or previous PartInstance if they are the same_?

Reply:
Its not that the same PartInstanceId will be listed multiple times, but multiple PartInstances with the same PartId. This was not directly relevant for this task, but is needed to allow having the same part in both current and next.

I have simplified this block as it was being a bit overly verbose, which hopefully makes it a bit easier to follow. The map is `Map<PartId, PartInstance>()`, so it was always checking PartId, and doing some additional checks for the special case